crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -277,19 +285,43 @@ impl RecursorDnsHandle {
         Error::recursion_exceeded(self.recursion_limit, depth, &query_name)?;
 
         let mut cname_chain = vec![];
+
         for rec in lookup.records().iter() {
             let CNAME(name) = rec.data() else {
                 continue;
             };
 
             let cname_query = Query::query(name.0.clone(), query_type);
 
+            // Block to drop the mutex guard after checking against MAX_CNAME_LOOKUPS
+            {
+                let mut count = cname_limit.lock();
+                *count += 1;
+
+                if *count > MAX_CNAME_LOOKUPS {
+                    warn!("cname limit exceeded for query {query}");
+                    return Err(ErrorKind::Proto(
+                        ProtoErrorKind::NotAllRecordsWritten {
```

Comment:
Should we use a new error type for this?

Reply:
I added a MaxRecordLimitExceeded type for this.